File: app/src/main/java/com/example/client/app/MyFirebaseService.java
package com.example.client.app;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.graphics.BitmapFactory;
import android.media.RingtoneManager;
import android.net.Uri;
import android.os.Build;
import android.util.Log;

import androidx.core.app.NotificationCompat;

import com.example.client.R;
import com.example.client.models.profile.ProfileModel;
import com.example.client.screens.noti.activity.NotificationActivity;
import com.google.firebase.messaging.FirebaseMessaging;
import com.google.firebase.messaging.FirebaseMessagingService;
import com.google.firebase.messaging.RemoteMessage;

import java.util.Map;
import java.util.Random;

import io.reactivex.Observable;

public class MyFirebaseService extends FirebaseMessagingService {
    private static final String TAG = "MyFirebaseService";
    @Override
    public void onMessageReceived(RemoteMessage remoteMessage) {
        // handle a notification payload.
        Log.d(TAG, "From: " + remoteMessage.getFrom());

        sendNotification(remoteMessage.getData());
    }

    @Override
    public void onNewToken(String token) {
        super.onNewToken(token);
        Log.d(TAG, "Refreshed token: " + token);

        sendRegistrationToServer(token);
    }

    private void sendRegistrationToServer(String token) {
        Preferences.newInstance().setDeviceToken(token);
        // TODO: Implement this method to send token to your app server.
    }

    private void sendNotification(Map data) {

        ProfileModel user = Preferences.newInstance().getProfile();
        PendingIntent pendingIntent = null;
        if(user == null){
            Intent intent = getPackageManager().getLaunchIntentForPackage("com.example.client");
            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
            pendingIntent = PendingIntent.getActivity(this, 0, intent, PendingIntent.FLAG_ONE_SHOT);
        }
        else {
            Intent intent = new Intent(this, NotificationActivity.class);
            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
            pendingIntent = PendingIntent.getActivity(this, 0, intent, PendingIntent.FLAG_ONE_SHOT);
        }
        String action = (String) data.get("action");
        String description = (String) data.get("description");

        String channelId = getString(R.string.app_name);
        Uri defaultSoundUri = RingtoneManager.getDefaultUri(RingtoneManager.TYPE_NOTIFICATION);

        NotificationCompat.Builder notificationBuilder =
                null;
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.N) {
            notificationBuilder = new NotificationCompat.Builder(this, channelId)
                    .setSmallIcon(R.drawable.ic_noti)
                    .setLargeIcon(BitmapFactory.decodeResource(getResources(), R.drawable.ic_noti))
                    .setContentTitle(action)
                    .setContentText(description)
                    .setAutoCancel(true)
                    .setSound(defaultSoundUri)
                    .setContentIntent(pendingIntent)
                    .setDefaults(Notification.DEFAULT_ALL)
                    .setPriority(NotificationManager.IMPORTANCE_HIGH);
        }

        NotificationManager notificationManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);

        // Since android Oreo notification channel is needed.
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel channel = new NotificationChannel(
                    channelId,
                    "Channel human readable title",
                    NotificationManager.IMPORTANCE_DEFAULT);

            notificationManager.createNotificationChannel(channel);
        }

        notificationManager.notify(new Random().nextInt(), notificationBuilder.build());
    }

    public Observable<Integer> resetToken(){
        return Observable.just(1).doOnNext(o -> FirebaseMessaging.getInstance().getToken());
    }

    public void deleteToken(){
        FirebaseMessaging.getInstance().deleteToken();
    }
}

File: app/src/main/java/com/example/client/app/MyFirebaseService.kt
package com.example.client.app

import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Intent
import android.graphics.Typeface
import android.media.RingtoneManager
import android.os.Build
import android.text.Spannable
import android.text.SpannableString
import android.text.style.StyleSpan
import android.util.Log
import androidx.core.app.NotificationCompat
import com.example.client.R
import com.example.client.models.notify.NotifyModel
import com.example.client.models.notify.getAction
import com.example.client.screens.notify.NotifyActivity
import com.example.client.screens.order.detail.OrderDetailActivity
import com.example.client.usecase.ProfileUseCase
import com.google.firebase.messaging.FirebaseMessaging
import com.google.firebase.messaging.FirebaseMessagingService
import com.google.firebase.messaging.RemoteMessage
import com.google.gson.Gson
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers

class MyFirebaseService : FirebaseMessagingService() {
    private val profileUseCase by lazy { ProfileUseCase.newInstance() }
    private val preferences by lazy { Preferences.newInstance() }
    private val firebaseMessaging by lazy { FirebaseMessaging.getInstance() }
    private val compositeDisposable = CompositeDisposable()

    override fun onMessageReceived(remoteMessage: RemoteMessage) {
        Log.d(TAG, "From: " + remoteMessage.from + ": " + Gson().toJson(remoteMessage.data))
        sendNotification(remoteMessage.data)
    }

    override fun onNewToken(token: String) {
        super.onNewToken(token)
        Log.d(TAG, "Refreshed token: $token")
        sendRegistrationToServer(token)
    }

    private fun sendRegistrationToServer(token: String) {
        preferences.deviceToken = token
        compositeDisposable.add(profileUseCase.updateDeviceToken(mEmail ?: return, device_token = token)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    if (it.is_error) {
                        firebaseMessaging.token
                    }
                }, {
                    it.printStackTrace()
                    firebaseMessaging.token
                }))
    }

    private fun sendNotification(data: Map<*, *>) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val user = Preferences.newInstance().profile
            val notify = NotifyModel(
                action = data["action"] as String,
                description = data["description"] as String,
                order_code = data["order"] as String?
            )

            val intent = user?.let {
                when(notify.action) {
                    Constants.NotifyAction.NOTIFY -> Intent(this, NotifyActivity::class.java)
                    Constants.NotifyAction.RECHARGE_SUCCESS -> packageManager.getLaunchIntentForPackage("com.example.client")
                    else -> Intent(this, OrderDetailActivity::class.java).apply {
                        putExtra(Constants.BundleKey.ORDER_CODE, notify.order_code)
                    }
                }
            } ?: kotlin.run {
                packageManager.getLaunchIntentForPackage("com.example.client")
            }

            intent?.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
            val pendingIntent = PendingIntent.getActivity(this, 0, intent, PendingIntent.FLAG_ONE_SHOT)
            val channelId = getString(R.string.app_name)
            val notificationManager = getSystemService(NOTIFICATION_SERVICE) as NotificationManager
            val title = SpannableString(getAction(notify)).apply {
                setSpan(StyleSpan(Typeface.BOLD), 0, getAction(notify).length, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
            }
            val notificationBuilder = NotificationCompat.Builder(this, channelId)
                    .setSmallIcon(R.drawable.logo_square)
                    .setContentTitle(title)
                    .setContentText(notify.description)
                    .setAutoCancel(true)
                    .setSound(RingtoneManager.getDefaultUri(RingtoneManager.TYPE_NOTIFICATION))
                    .setContentIntent(pendingIntent)
                    .setFullScreenIntent(pendingIntent, true)
                    .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                    .setPriority(NotificationManager.IMPORTANCE_HIGH)
                    .setStyle(NotificationCompat.BigTextStyle().bigText(notify.description))
            val channel = NotificationChannel(
                    channelId,
                    "CSC channel",
                    NotificationManager.IMPORTANCE_HIGH)
            notificationManager.createNotificationChannel(channel)
            notificationManager.notify(System.currentTimeMillis().toInt(), notificationBuilder.build())
        }
    }

    fun resetToken(email: String) {
        mEmail = email
        firebaseMessaging.token
    }

    fun deleteToken() {
        firebaseMessaging.deleteToken()
    }

    companion object {
        private const val TAG = "MyFirebaseService"
        private var mEmail: String? = null
    }

    override fun onDestroy() {
        super.onDestroy()
        compositeDisposable.clear()
    }
}